SourceKitStressTester/Sources/StressTester/RequestDurationManager.swift

```diff
@@ -0,0 +1,99 @@
+  /// Creates a logarithmic histogram. For each `key`, the `value` contains the
+  /// number of elements in this array that are smaller than `2 ^ key` but not
+  /// smaller than `2 ^ (key - 1)`.
+  var logHistogram: [Int: Int] {
+    var result: [Int: Int] = [:]
+    for value in self {
+      let log = Int(log2(Double(value))) + 1
+      result[log, default: 0] += 1
+    }
+    return result
+  }
+
+  /// The average value in this array, rounded to an integer or 0 if the array
+  /// is empty
+  var average: Int {
+    if self.isEmpty {
+      return 0
+    } else {
+      return self.reduce(0, { $0 + $1 }) / self.count
+    }
+  }
+}
+
+/// Captures aggregated information about executing a certain request kind on a
+/// certain file.
+struct AggregatedRequestDurations: Codable {
+  /// The log histogram created by the extension on `Array` above
+  let logHistogram: [Int: Int]
+
+  /// The average number of instructions a request took to execute
+  let average: Int
+
+  /// The total number of requests of a given kind executed for a given file
+  let requestsExecuted: Int
+
+  init(instructionCounts: [Int]) {
+    logHistogram = instructionCounts.logHistogram
+    average = instructionCounts.average
+    requestsExecuted = instructionCounts.count
+  }
+}
+
+/// Contains aggregated information for all request kinds of all files executed
+/// by the stress tester.
+fileprivate struct RequestDurations: Codable {
+  // FIXME: We should be using `RequestKind` as the inner key but that causes
+  // the inner dictionary to be serialized as an array (rdar://78099769)
+  /// Maps file paths to request kinds to aggregated request duration information
+  var files: [String: [String: AggregatedRequestDurations]]
+}
+
+/// Collects the durations that requests executed by the stress tester took and
+/// writes them to `jsonFile` where the durations are collected together with
+/// all other stress tester runs.
+/// We are aggregating the results early (just keeping track of the average for
+/// each request type and a logarithmic histogram because keeping track of all
+/// request durations would result in a JSON file that is too large to handle
+/// easily.
+class RequestDurationManager {
+  /// The file that stores the request durations and that gets updated as new
+  /// aggreaged information is added
+  let jsonFile: URL
+
+  init(jsonFile: URL) {
+    self.jsonFile = jsonFile
+  }
+
+  private func getRequestDurations() throws -> RequestDurations {
+    guard FileManager.default.fileExists(atPath: jsonFile.path) else {
+      return RequestDurations(files: [:])
+    }
+    let data = try Data(contentsOf: jsonFile)
+    return try JSONDecoder().decode(RequestDurations.self, from: data)
+  }
+
+  func add(aggregatedDurations: AggregatedRequestDurations, for file: String, requestKind: RequestKind) throws {
+    // FIXME: We are racing for the timings file here if two stress tester invocations try to update it simultaneously
```

Comment:
I trust the `FileManager` to get a lock on the file so that it will always decode fine. But we might be dropping information here:
- Process A reads the file
- Process B reads the file
- Process A adds its information in-process
- Process B adds its information in-process
- Process A writes its updated file
- Process B writes its updated file

In this case we’d be dropping the information from Process A. I think, the correct way of doing it would be to obtain a file handle to the file for reading and writing so that it’s locked but that would mean that we couldn’t use all the convenience methods. And since we already use the racing paradigm when writing out the expected issues, I think it’s not too bad in practice.

Reply:
One difference here is that I'd expect we wouldn't be updating the issues as often as we'd be updating the instruction counts (ie. not every process will have issues but every process *will* have instruction counts). I'm happy for you to leave this for now though.

Reply:
I spent way too long figuring how to lock a file only to discover in the end that tools-support-core has a method for that. We should be locking the file properly now, avoiding any race conditions.